### PR TITLE
fix: Allow newer pydantic versions.

### PR DESCRIPTION
**Pull request by yurnih**
The current version of pydantic (v1.6.2) does not fully support python 3.9, mainly when using self-referenced type annotations. Although it doesn't show up as failure in CI, it could be a possible failure when integrating geolib with other projects. I changed tilde constraint to caret to allow later versions of pydantic, but this could be lead to other failure at a latter stage as happened earlier (it is also possible to limit the upper version, like `<1.9`).

The upgrade caused some validators to fail due to `validate_assignment = True`, where some properties are not yet defined, so the order of defining those properties is changed.

Furthermore, some Config-options are removed since they all inherit https://github.com/Deltares/GEOLib/blob/54de109460a836caed51f0f26974a51f39edcfa3/geolib/models/base_model_structure.py#L12-L19.

### CODE REVIEW

File: geolib/models/dstability/internal.py [hunk 1]
```diff
@@ -1866,6 +1894,13 @@ def add_default_stage(self, label: str, notes: str, unique_start_id=500) -> int:
         ]
         self.states += [State(Id=str(unique_start_id + 3))]
         self.statecorrelations += [StateCorrelation(Id=str(unique_start_id + 4))]
+        self.soillayers += [SoilLayerCollection(Id=str(unique_start_id + 5))]
```
Comment (evetion) [inline geolib/models/dstability/internal.py:1897]: These changes are probably great, but don't belong in this PR, right?

Comment (yurnih) [reply geolib/models/dstability/internal.py:1897]: I discovered by testing that the order of assignment is important in later verions of pydantic.

The block of;
https://github.com/Deltares/GEOLib/blob/d0b9a74e59111a76cc4bb2b993e00fade9bbf1e7/geolib/models/dstability/internal.py#L1904-L1920

Triggers the validation at https://github.com/Deltares/GEOLib/blob/d0b9a74e59111a76cc4bb2b993e00fade9bbf1e7/geolib/models/dstability/internal.py#L1771-L1772

However, when I use the old order here, the setting of `skip_on_failure` got ignored when using `validate_assignment` and thus triggering an error in tests. By adding the sub items first, and afterwards the verification of foreign-keys, this error doesn't show up.
